### PR TITLE
Extract in memory parsing helper for settings

### LINKED ISSUE #696: [Epic] Phase 2: Decouple Settings Parsing
---
name: Feature Request
about: Suggest an idea for this project
title: '[FEATURE] Epic: Phase 2 - Decouple Settings Parsing'
labels: enhancement, architecture
assignees: ''

---

**Description**:

Tracking issue for completely isolating `Settings` data parsing from disk I/O to enable pure in-memory testing.

**Proposed Implementation**:

- [x] Sub-task 2.1: Extract `apply_config_to_input_map` helper.
- [x] Sub-task 2.2: Refactor `load_input_mappings` to delegate to helper.
- [x] Sub-task 2.3: Refactor GUT tests (`test_settings_ec.gd`).
- [x] Sub-task 2.4: Refactor GdUnit4 tests (`test_settings.gd`).

---

### **[Epic] Phase 2: Decouple Settings Parsing (#696)**

| Goal / Acceptance Criteria | Target Issue | Target File(s) | Implementation Strategy |
| :--- | :--- | :--- | :--- |
| **Extract `apply_config_to_input_map` helper** | #697 | `settings.gd` | Create a pure-GDScript helper that accepts an in-memory `ConfigFile` and maps it to `InputMap`. |
| **Refactor `load_input_mappings`** | #698 | `settings.gd` | Strip parsing logic from the main loader; restrict it to file I/O and delegate the rest to the new helper. |
| **Update `test_settings_ec.gd` (GUT)** | #699 | `test_settings_ec.gd` | Remove file saving from corruption tests. Instantiate a `ConfigFile`, inject garbage data, and pass it to the helper. *(Ensures zero C++ native errors)*. |
| **Update `test_settings.gd` (GdUnit4)** | #700 | `test_settings.gd` | Rewrite GdUnit4 error-handling tests to use the exact same in-memory `ConfigFile` injection pattern. *(Ensures zero C++ native errors)*. |

---

> **Note:** If you need any additional information or context regarding this refactor, please refer to the original issue here: https://github.com/ikostan/SkyLockAssault/issues/589

### LINKED ISSUE #697: [FEATURE] Extract in-memory parsing helper for Settings
---
name: Feature Request
about: Suggest an idea for this project
title: '[FEATURE] Extract in-memory parsing helper for Settings'
labels: enhancement, architecture
assignees: ''
---

**Description**:

Create a pure-GDScript helper function for parsing input mappings from a `ConfigFile`.

**Why is this useful?** Step one in decoupling parsing from disk I/O.

**Proposed Implementation**:

* **Dev Task**: Create `apply_config_to_input_map(config: ConfigFile, actions: Array[String])` in `settings.gd`. Move the deserialization and InputMap mapping logic here.
* **Documentation Task**: Add comments explicitly stating this is for in-memory processing and testing.

---

> **Note:** If you need any additional information or context regarding this refactor, please refer to the original issue here: https://github.com/ikostan/SkyLockAssault/issues/589

### LINKED ISSUE #698: [FEATURE] Refactor load_input_mappings to use new parsing helper
---
name: Feature Request
about: Suggest an idea for this project
title: '[FEATURE] Refactor load_input_mappings to use new parsing helper'
labels: enhancement, architecture
assignees: ''
---

**Description**:

Update the main loader to act strictly as an I/O coordinator.

**Why is this useful?** Enforces Separation of Concerns.

**Proposed Implementation**:

* **Dev Task**: Strip parsing logic from `Settings.load_input_mappings()`. It should now only read the file (via `safe_load_config`) and pass the `ConfigFile` object to `apply_config_to_input_map`.

---

> **Note:** If you need any additional information or context regarding this refactor, please refer to the original issue here: https://github.com/ikostan/SkyLockAssault/issues/589

### LINKED ISSUE #699: [FEATURE] Refactor GUT input settings tests to use in-memory ConfigFile
name: Feature Request
about: Suggest an idea for this project
title: '[FEATURE] Refactor GUT input settings tests to use in-memory ConfigFile'
labels: testing, enhancement
assignees: ''

Description:
Refactor GUT edge-case tests in `test_settings_ec.gd` to test in-memory configuration parsing directly without unnecessary disk I/O.

Why is this useful? 
Bypasses file system overhead and avoids C++ encryption/decryption routines during in-memory data validation tests.

Proposed Implementation:
* Testing Task: Update parsing edge-case tests (`test_ec_04`, `test_ec_05`, `test_ec_07`, `test_ec_10`) to construct in-memory `ConfigFile` instances and pass them directly to `Settings.apply_config_to_input_map()`.
* Isolation Task: Reserve disk saving/loading strictly for true file system tests (like `test_ec_06_save_fails_gracefully` and `test_ec_09_last_input_device_validation`).

---

> **Note:** If you need any additional information or context regarding this refactor, please refer to the original issue here: https://github.com/ikostan/SkyLockAssault/issues/589

### LINKED ISSUE #700: [FEATURE] Refactor GdUnit4 input settings tests to use in-memory ConfigFile
---
name: Feature Request
about: Suggest an idea for this project
title: '[FEATURE] Refactor GdUnit4 input settings tests to use in-memory ConfigFile'
labels: testing, enhancement
assignees: ''
---

**Description**:

Update `test_settings.gd` to test error handling without saving files to disk.

**Why is this useful?** Standardizes testing methodology across frameworks and prevents engine crashes.

**Proposed Implementation**:

* **Testing Task**: Rewrite tests (like `test_load_error_handling`) to use in-memory `ConfigFile` injection with `apply_config_to_input_map`.

---

> **Note:** If you need any additional information or context regarding this refactor, please refer to the original issue here: https://github.com/ikostan/SkyLockAssault/issues/589

### PR DESCRIPTION
**Pull request by ikostan**
---
name: Default Pull Request Template
about: Suggesting changes to SkyLockAssault
title: ''
labels: ''
assignees: ''
---

**PR #841 Summary: Extract in-memory parsing helper for settings**

### Overview

This PR by **@ikostan** refactors the input settings system in SkyLockAssault (Godot-based). It extracts parsing logic into a reusable in-memory helper, improves test coverage/isolation, hardens legacy migration and error handling, and updates several GitHub Actions in CI workflows.

### Key Changes

- **New Features**:
  - Introduced `apply_config_to_input_map()` helper for applying `ConfigFile` input mappings directly to Godot's `InputMap` (pure in-memory, no side effects).
  - Added test-only helpers (`backfill_missing_defaults`, `set_needs_save_for_test`, `needs_save()`) for better test control.

- **Improvements & Bug Fixes**:
  - Better handling of corrupt/unsupported config values while preserving existing bindings.
  - Improved legacy plaintext migration and `_needs_save` flag logic.
  - Refactored `load_input_mappings()` to delegate parsing and clarify responsibilities.
  - Enhanced conflict resolution and normalization of input events.

- **Testing**:
  - Reorganized tests to separate disk I/O from in-memory parsing behavior.
  - Expanded coverage for edge cases (legacy formats, unbound actions, conflicts).

- **CI/CD**:
  - Updated GitHub Actions to newer versions (setup-python, setup-node, markdownlint, release-drafter, CodeQL SARIF upload, etc.).

### Impact

- Cleaner, more testable settings code.
- Reduced coupling between file I/O and parsing logic.
- Improved maintainability and robustness for input configuration.

---

## Related Issue

Closes #ISSUE_NUMBER (if applicable)

## Changes

- [x] List key changes here (e.g., "Updated Jump.gd to use Godot 4.4's new Tween
  system")
- [x] Any breaking changes? (e.g., "Deprecated old signal; migrate to new one")

## Testing

- [x] Ran the game in Godot v4.5 editor—describe what you tested (e.g., "Jump
  works on Win10 with 60 FPS")
- [x] Any new unit tests added? (Link to test scene if yes)
- [x] Screenshots/GIFs if UI-related: (Attach below)

## Checklist

- [x] Code follows Godot style guide (e.g., snake_case for variables)
- [x] No console errors in editor/output
- [x] Ready for review!

## Additional Notes

Anything else? (e.g., "Tested on Win10 64-bit; needs Linux validation")

## Summary by Sourcery

Extract input settings parsing into a reusable in-memory helper and align tests and workflows with the new API.

New Features:
- Introduce apply_config_to_input_map helper for applying ConfigFile input mappings directly to Godot's InputMap.
- Expose test-only helpers to backfill missing defaults and control/inspect the internal needs_save flag.

Bug Fixes:
- Ensure corrupt or unsupported config values are safely ignored while preserving existing input bindings.
- Guarantee legacy plaintext migration sets and clears the save-needed flag only for old formats.
- Maintain correct handling of legacy key formats and explicit unbound actions when using the new parser.

Enhancements:
- Refactor load_input_mappings to delegate parsing to the new in-memory helper and clarify its responsibilities.
- Reorganize and extend settings tests to distinguish disk I/O persistence from pure in-memory parsing behavior.
- Improve event label and pause binding tests for clearer expectations and reduced noise.

CI:
- Update GitHub Actions workflows to newer pinned versions of setup-python, setup-node, markdownlint, release-drafter, and CodeQL upload-sarif actions.

## Summary by CodeRabbit

* **Chores**
  * Updated CI and automation actions to newer revisions, including Python/Node setup for browser/CI runs.
  * Refreshed markdown lint, release drafting, security scanning, and SARIF upload action revisions while preserving existing workflow behavior.
  * Refined input-mapping configuration handling and improved conflict/migration application flow.
  * Expanded/adjusted unit and integration tests to validate parsing, backfilling, and conflict behavior using in-memory configuration instead of disk fixtures.

---

**Bots/AI Contributors to PR #841**

This PR includes significant automation and AI-assisted contributions alongside human work.

### Bot/AI Contributions

- **@dependabot[bot]**: Multiple dependency update commits for GitHub Actions (e.g., `actions/setup-node`, `actions/setup-python`, `release-drafter/release-drafter`, `github/codeql-action/upload-sarif`, `DavidAnson/markdownlint-cli2-action`). These refreshed CI workflows to newer pinned versions.
- **@sourcery-ai**: Provided a detailed PR summary highlighting new features (e.g., `apply_config_to_input_map` helper), bug fixes, enhancements, test improvements, and CI updates.
- **@coderabbitai**: Offered a summary focusing on chores (CI/automation updates), refactoring of input-mapping handling, and expanded tests for parsing/backfilling.
- **@Copilot**: Collaborated on at least one commit involving refactoring input config normalization logic.

(DeepSourceReview / @deepsource-io was not visibly active in the commits, summaries, or reviews for this specific PR based on available details, though it is integrated in the repository.)

### @ikostan Contributions

@ikostan drove the core changes as the primary author:
- Refactored input settings loading by extracting a reusable in-memory parsing helper (`apply_config_to_input_map` and supporting methods like `_normalize_input_value`).
- Improved test isolation (disk I/O vs. pure in-memory parsing), added test-only helpers, hardened legacy migration/corner-case handling, and updated documentation/sequence diagrams.
- Performed merges of Dependabot PRs and various follow-up refinements to settings.gd, tests, and workflows.

This structure ensures bots/AI are properly credited in GitHub's contributors list while separating human effort.

### CODE REVIEW

File: .github/workflows/yamllint.yml [hunk 1]
```diff
@@ -34,7 +34,7 @@ jobs:
         # yamllint disable rule:line-length
       - name: Set up Python ${{ matrix.python-version }}  # yamllint disable-line rule:quoted-strings
         # yamllint enable rule:line-length
-        uses: "actions/setup-python@v6"
+        uses: "actions/setup-python@v7"
```
[the remark this answers is not in the document]
Comment (ikostan) [reply .github/workflows/yamllint.yml:37]: Actually, v7.0.0 was released 2 days ago!

**Comment by ikostan** [comment]
@sourcery-ai review

**Comment by ikostan** [comment]
> .github/workflows/browser_test.yml (1)
> > `75-77`: _📐 Maintainability & Code Quality_ | _🔵 Trivial_ | _⚡ Quick win_
> > **Remove the stale setup-python version comments.**
> > Line 77 now uses v7, but the surrounding comments still say the workflows use v6 and instruct updating from v5 to v6.

---

Fixed

**Comment by ikostan** [comment]
> * Using `apply_config_to_input_map(ConfigFile.new())` in `gut/test_settings_ec.gd:before_each` will call `_add_missing_defaults` and potentially flip `_needs_save` based on an empty in-memory config; consider either resetting `_needs_save` around this call or adding a parameter to skip defaults/migration when using the helper purely for tests.
> * `apply_config_to_input_map` mutates `_needs_save` and runs `_add_missing_defaults` even when used for targeted, in-memory updates (e.g., some EC tests); if you intend this helper to be usable for partial/ephemeral updates, it may be worth separating the pure mapping logic from default-backfilling/migration, or adding a flag to control those side

---

Great catch and spot-on architectural feedback! 🎯

We have addressed this by strictly decoupling pure mapping from persistence side effects:

1. **Made `apply_config_to_input_map` Pure:** Removed `_add_missing_defaults()` execution and `_needs_save` state mutations from `apply_config_to_input_map()`. It now acts purely as an in-memory ConfigFile-to-InputMap deserializer without side effects.
2. **I/O Coordinator Role:** Moved default backfilling (`_add_missing_defaults()`) and `_needs_save` state flagging into `load_input_mappings()`, keeping disk I/O and persistence updates contained.
3. **Test Suite Isolation:** Updated unit tests in `test_settings_ec.gd` and `test_settings.gd` so tests explicitly invoke `_add_missing_defaults()` only when testing the default-fallback chain, keeping state strictly isolated.

**Comment by ikostan** [comment]
@sourcery-ai review

File: scripts/core/settings.gd [hunk 2]
```diff
@@ -799,3 +757,67 @@ func load_last_input_device() -> void:
 		Globals.current_input_device = device if device in ["keyboard", "gamepad"] else "keyboard"
 	else:
 		Globals.current_input_device = "keyboard"
+
+
+## Applies input mapping definitions from an in-memory ConfigFile directly into Godot's InputMap.
```
[the remark this answers is not in the document]
Comment (ikostan) [reply scripts/core/settings.gd:762]: Good eye! 🎯

Updated `apply_config_to_input_map()` to eliminate all direct mutations of `_needs_save`:

1. **Returned Conflict Status:** `apply_config_to_input_map()` now returns a `bool` indicating whether conflict resolution took place, leaving zero side-effect mutations on singleton state.
2. **Coordinator Handling:** `load_input_mappings()` collects the boolean return value from `apply_config_to_input_map()` alongside `_add_missing_defaults()` to determine if `_needs_save` should be flagged.

File: scripts/core/settings.gd [hunk 3]
```diff
@@ -799,3 +757,67 @@ func load_last_input_device() -> void:
...
+						Globals.log_message(
+							"Non-string item in array for action '" + action + "': skipped",
+							Globals.LogLevel.WARNING
+						)
+			elif value is String:
+				serialized_events = [value]
+			elif value is int:
+				serialized_events = ["key:" + str(value)]
+
+			InputMap.action_erase_events(action)
```
[the remark this answers is not in the document]
Comment (ikostan) [reply scripts/core/settings.gd]: Resolved

**Comment by ikostan** [comment]
> * The new `apply_config_to_input_map` helper is documented as a pure, in‑memory parser but still mutates `_needs_save` when resolving conflicts; consider either adjusting the documentation to call out this side effect or splitting the conflict/migration bookkeeping into a separate wrapper so the helper can be used without hidden global state changes.
> * Several tests now invoke private/internal APIs like `Settings._add_missing_defaults` and poke `_needs_save` directly; if possible, consider adding small public or test-only helpers for these behaviors so tests are less coupled to the singleton’s internal implementation details.

---

All points addressed! 🎯

1. **Pure Mapping Function:** `apply_config_to_input_map()` no longer touches `_needs_save`[cite: 15]. It now returns a `bool` indicating if conflicts were resolved[cite: 15]. `load_input_mappings()` handles all `_needs_save` aggregation[cite: 15].
2. **Public Test API Decoupling:** Added clean public wrappers in `settings.gd` (`backfill_missing_defaults()`, `set_needs_save_for_test()`, and `needs_save()`) so unit tests no longer poke private members or internal functions directly[cite: 15].

**Comment by ikostan** [comment]
> scripts/core/settings.gd (1)
> > `625-632`: _📐 Maintainability & Code Quality_ | _🔵 Trivial_ | _💤 Low value_
> > **Duplicate docstring line.**
> > Line 626 repeats line 625 verbatim.
> > 
> > ✏️ Proposed fix
> > ```diff
> >  ## Loads input mappings from config, overriding project defaults only if saved.
> > -## Loads input mappings from config, overriding project defaults only if saved.
> >  ## Acts strictly as an I/O coordinator by loading the file via `Globals.safe_load_config`,
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 
> > 🤖 Prompt for AI Agents

---

Fixed

**Comment by ikostan** [comment]
> test/gut/test_settings_ec.gd (1)
> > `57-77`: _🎯 Functional Correctness_ | _🔵 Trivial_ | _⚡ Quick win_
> > **Test uses a keycode that coincides with the actual default, weakening the assertion.**
> > `TEST_ACTION` ("speed_up") is set to old-format int `87`, which is the same physical keycode as `DEFAULT_KEYBOARD["speed_up"]` (`KEY_W`). Since `before_each()` already backfills `speed_up` to `KEY_W` before this test runs, the assertion at line 72-74 would pass even if the int-deserialization branch in `apply_config_to_input_map` were broken, because the default-backfill path would produce the same keycode. Use a non-default keycode (e.g. one of the suite's `TEST_KEY_*` constants) to make the assertion actually exercise the migration path.

---

Great catch! 🎯

Updated `test_ec_04_legacy_mixed_formats` in `test_settings_ec.gd` to use keycode `89` (`KEY_Y`) instead of `87` (`KEY_W`). 

This distinguishes the migrated legacy integer mapping from the default key assignment, ensuring the assertion explicitly validates integer deserialization without false positives from the default backfill step.

**Comment by ikostan** [comment]
@sourcery-ai review

File: scripts/core/settings.gd [hunk 4]
```diff
@@ -799,3 +757,97 @@ func load_last_input_device() -> void:
...
+						Globals.log_message(
+							"Non-string item in array for action '" + action + "': skipped",
+							Globals.LogLevel.WARNING
+						)
+			elif value is String:
+				serialized_events = [value]
+			elif value is int:
+				serialized_events = ["key:" + str(value)]
+			else:
+				Globals.log_message(
```
[the remark this answers is not in the document]
Comment (ikostan) [reply scripts/core/settings.gd]: Great catch! 🎯

Addressed this by short-circuiting unsupported value types before event erasure:

1. **Preserved Existing Bindings:** Added a `continue` statement within the `else` branch of `apply_config_to_input_map()`. If an unsupported value type (e.g., `bool`, `Dictionary`, `float`) is encountered, a warning is logged and execution immediately skips to the next action, keeping existing `InputMap` bindings intact instead of clearing them.
2. **Added Unit Test Coverage:** Added `test_ec_15_unsupported_type_preserves_existing_events` in `test_settings_ec.gd` to explicitly verify that invalid config types preserve pre-existing bindings.

**Comment by ikostan** [comment]
> * The new apply_config_to_input_map helper largely duplicates the previous inline logic in load_input_mappings; consider extracting common sub-steps (e.g., serialization normalization and conflict handling) into smaller private helpers to reduce future maintenance overhead.
> * Now that backfill_missing_defaults and set_needs_save_for_test/needs_save are public wrappers, it may be worth marking them clearly as test-only or moving them into a dedicated test support module to avoid accidental use in production code.
> * In test_settings_ec.gd, EC-05 still calls the private _add_missing_defaults directly while other tests use backfill_missing_defaults; aligning all tests to the public wrapper would keep the suite consistent and avoid relying on internal implementation details.

---

All 3 points addressed! 🎯

1. **Extracted Normalization Helper:** Extracted array/string/int value parsing into a private helper function (`_normalize_input_value`), cleanly separating parsing sub-steps within `apply_config_to_input_map()`.
2. **Explicit `[TEST ONLY]` Annotations:** Annotated `backfill_missing_defaults()`, `set_needs_save_for_test()`, and `needs_save()` docstrings with explicit `[TEST ONLY]` tags to prevent accidental production usage.
3. **Aligned Test Suite Wrapper Calls:** Updated `test_ec_05_corrupt_parse_error` in `test_settings_ec.gd` to use the public `backfill_missing_defaults()` wrapper consistently instead of calling private internal methods directly.

**Comment by ikostan** [comment]
@sourcery-ai review

**Comment by ikostan** [comment]
> * In `test_ec_05_corrupt_parse_error` you still call the private `Settings._add_missing_defaults`; for consistency with the new test-only API, consider switching this to `Settings.backfill_missing_defaults(...)` so tests no longer depend on private internals.
> * The new test-only helpers (`backfill_missing_defaults`, `set_needs_save_for_test`, `needs_save`) are now part of the public `Settings` API; if possible, consider moving them into a dedicated test helper script or making them conditional to avoid expanding the production-facing surface with test-only methods.

---

Thanks for pointing that out! 🎯

1. **Test Wrapper Alignment:** Updated `test_ec_05_corrupt_parse_error` in `test_settings_ec.gd` to call `Settings.backfill_missing_defaults(cfg)` instead of calling the internal `_add_missing_defaults` method directly.
2. **Public Surface Boundary:** The test helpers (`backfill_missing_defaults`, `set_needs_save_for_test`, `needs_save`) are explicitly marked with `## [TEST ONLY]` docstrings to clearly mark them for test suite initialization without introducing complex conditional loading or test script inheritance to the autoload singleton.

**Comment by ikostan** [comment]
@sourcery-ai guide
